desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        grid=None,
+        field_grid=None,
+        name="Vacuum boundary error",
+    ):
+        if target is None and bounds is None:
+            target = 0
+        self._grid = grid
+        self._ext_field = ext_field
+        self._field_grid = field_grid
+        things = [eq]
+
+        super().__init__(
+            things=things,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        if self._grid is None:
+            grid = LinearGrid(
+                rho=np.array([1.0]),
+                M=eq.M_grid,
+                N=eq.N_grid,
+                NFP=int(eq.NFP),
+                sym=False,
+            )
+        else:
+            grid = self._grid
+
+        pres = np.max(np.abs(eq.compute("p")["p"]))
+        curr = np.max(np.abs(eq.compute("current")["current"]))
+        errorif(
+            not np.all(grid.nodes[:, 0] == 1.0),
+            ValueError,
+            "grid contains nodes not on rho=1",
+        )
+        warnif(
+            pres > 1e-8,
+            UserWarning,
+            f"Pressure is non-zero (max {pres} Pa), "
+            + "VacuumBoundaryError will be incorrect.",
+        )
+        warnif(
+            curr > 1e-8,
+            UserWarning,
+            f"Current is non-zero (max {curr} A), "
+            + "VacuumBoundaryError will be incorrect.",
+        )
+
+        self._eq_data_keys = [
+            "B",
+            "R",
+            "zeta",
+            "Z",
+            "n_rho",
+            "|e_theta x e_zeta|",
+        ]
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        profiles = get_profiles(self._eq_data_keys, obj=eq, grid=grid)
+        transforms = get_transforms(self._eq_data_keys, obj=eq, grid=grid)
+
+        self._constants = {
+            "transforms": transforms,
+            "profiles": profiles,
+            "ext_field": self._ext_field,
+            "quad_weights": np.sqrt(np.tile(transforms["grid"].weights, 2)),
+        }
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        self._dim_f = 2 * grid.num_nodes
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["B"] * scales["R0"] * scales["a"]
```

Comment:
shouldn't the normalization be this for the first half of the returned value Bn, but should have a `scales["B"]**2` for the 2nd half where it is the difference between the squared B magnitudes?

Reply:
Yeah wasn't sure how to handle this. As it currently stands `normalization` is a single scalar for each objective, assuming that each objective only returns one thing. But for the free boundary stuff its super wasteful to split the different outputs into different objectives.